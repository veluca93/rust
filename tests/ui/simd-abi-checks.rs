//@ only-x86_64
//@ build-fail

#![feature(avx512_target_feature)]
#![feature(portable_simd)]
#![allow(improper_ctypes_definitions)]

use std::arch::x86_64::*;

#[repr(transparent)]
struct Wrapper(__m256);

unsafe extern "C" fn w(_: Wrapper) {
    //~^ ABI error: this function definition uses a avx vector type, which is not enabled
    todo!()
}

unsafe extern "C" fn f(_: __m256) {
    //~^ ABI error: this function definition uses a avx vector type, which is not enabled
    todo!()
}

unsafe extern "C" fn g() -> __m256 {
    //~^ ABI error: this function definition uses a avx vector type, which is not enabled
    todo!()
}

#[target_feature(enable = "avx2")]
unsafe extern "C" fn favx(_: __m256) {
    todo!()
}

#[target_feature(enable = "avx")]
unsafe extern "C" fn gavx() -> __m256 {
    todo!()
}

fn as_f64x8(d: __m512d) -> std::simd::f64x8 {
    unsafe { std::mem::transmute(d) }
}

unsafe fn test() {
    let arg = std::mem::transmute([0.0f64; 8]);
    as_f64x8(arg);
}

fn main() {
    unsafe {
        f(g());
        //~^ ERROR ABI error: this function call uses a avx vector type, which is not enabled in the caller
        //~| ERROR ABI error: this function call uses a avx vector type, which is not enabled in the caller
    }

    unsafe {
        favx(gavx());
        //~^ ERROR ABI error: this function call uses a avx vector type, which is not enabled in the caller
        //~| ERROR ABI error: this function call uses a avx vector type, which is not enabled in the caller
    }

    unsafe {
        test();
    }

    unsafe {
        w(Wrapper(g()));
        //~^ ERROR ABI error: this function call uses a avx vector type, which is not enabled in the caller
        //~| ERROR ABI error: this function call uses a avx vector type, which is not enabled in the caller
    }
}
