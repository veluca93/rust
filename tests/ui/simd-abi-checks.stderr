error: ABI error: this function call uses a avx vector type, which is not enabled in the caller
  --> $DIR/simd-abi-checks.rs:49:11
   |
LL |         f(g());
   |           ^^^
   |
   = help: consider enabling it globally (-C target-feature=+avx) or locally (#[target_feature(enable="avx")])

error: ABI error: this function call uses a avx vector type, which is not enabled in the caller
  --> $DIR/simd-abi-checks.rs:49:9
   |
LL |         f(g());
   |         ^^^^^^
   |
   = help: consider enabling it globally (-C target-feature=+avx) or locally (#[target_feature(enable="avx")])

error: ABI error: this function call uses a avx vector type, which is not enabled in the caller
  --> $DIR/simd-abi-checks.rs:55:14
   |
LL |         favx(gavx());
   |              ^^^^^^
   |
   = help: consider enabling it globally (-C target-feature=+avx) or locally (#[target_feature(enable="avx")])

error: ABI error: this function call uses a avx vector type, which is not enabled in the caller
  --> $DIR/simd-abi-checks.rs:55:9
   |
LL |         favx(gavx());
   |         ^^^^^^^^^^^^
   |
   = help: consider enabling it globally (-C target-feature=+avx) or locally (#[target_feature(enable="avx")])

error: ABI error: this function call uses a avx vector type, which is not enabled in the caller
  --> $DIR/simd-abi-checks.rs:65:19
   |
LL |         w(Wrapper(g()));
   |                   ^^^
   |
   = help: consider enabling it globally (-C target-feature=+avx) or locally (#[target_feature(enable="avx")])

error: ABI error: this function call uses a avx vector type, which is not enabled in the caller
  --> $DIR/simd-abi-checks.rs:65:9
   |
LL |         w(Wrapper(g()));
   |         ^^^^^^^^^^^^^^^
   |
   = help: consider enabling it globally (-C target-feature=+avx) or locally (#[target_feature(enable="avx")])

error: ABI error: this function definition uses a avx vector type, which is not enabled
  --> $DIR/simd-abi-checks.rs:23:1
   |
LL | unsafe extern "C" fn g() -> __m256 {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider enabling it globally (-C target-feature=+avx) or locally (#[target_feature(enable="avx")])

error: ABI error: this function definition uses a avx vector type, which is not enabled
  --> $DIR/simd-abi-checks.rs:18:1
   |
LL | unsafe extern "C" fn f(_: __m256) {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider enabling it globally (-C target-feature=+avx) or locally (#[target_feature(enable="avx")])

error: ABI error: this function definition uses a avx vector type, which is not enabled
  --> $DIR/simd-abi-checks.rs:13:1
   |
LL | unsafe extern "C" fn w(_: Wrapper) {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider enabling it globally (-C target-feature=+avx) or locally (#[target_feature(enable="avx")])

error: aborting due to 9 previous errors

